Tests/NIOPosixTests/DatagramChannelTests.swift

```diff
@@ -1409,4 +1413,136 @@ class DatagramChannelTests: XCTestCase {
+        let allocator = FixedSizeRecvByteBufferAllocator(capacity: 1 << 16)
+        XCTAssertNoThrow(try receiver.setOption(ChannelOptions.recvAllocator, value: allocator).wait())
+
+        let buffer = self.firstChannel.allocator.buffer(repeating: 1, count: segmentSize * segments)
+        let writeData = AddressedEnvelope(remoteAddress: receiver.localAddress!, data: buffer)
+        XCTAssertNoThrow(try sender.writeAndFlush(NIOAny(writeData)).wait())
+
+        let received = try receiver.waitForDatagrams(count: 1)
+        let hasGoodGROSupport = received.first!.data.readableBytes == buffer.readableBytes
+
+        if !hasGoodGROSupport {
+            // Not well supported: check we receive enough datagrams of the expected size.
+            let datagrams = try receiver.waitForDatagrams(count: segments)
+            for datagram in datagrams {
+                XCTAssertEqual(datagram.data.readableBytes, segmentSize)
+            }
+
+            #if os(Linux)
+            let info = System.systemInfo
+            // If our kernel is more recent than 5.15 and we don't have good GRO support then
+            // something has gone wrong (or our assumptions about kernel support are incorrect).
+            if let major = info.release.major, let minor = info.release.minor {
+                if major >= 6 || (major == 5 && minor >= 15) {
+                    XCTFail("Platform does not have good GRO support: \(info.release.release)")
+                }
+            } else {
+                XCTFail("Unable to determine Linux x.y release from '\(info.release.release)'")
+            }
+            #endif
+        }
+
+        return hasGoodGROSupport
+    }
+}
+
+extension System {
+    #if os(Linux)
+    internal static let systemInfo: SystemInfo = {
+        var info = utsname()
+        assert(CNIOLinux_system_info(&info) == 0)
+        return SystemInfo(utsname: info)
+    }()
+
+    struct SystemInfo {
+        var machine: String
+        var nodeName: String
+        var sysName: String
+        var release: Release
+        var version: String
+
+        struct Release {
+            var release: String
+
+            var major: Int?
+            var minor: Int?
+
+            init(parsing release: String) {
+                self.release = release
+
+                let components = release.split(separator: ".", maxSplits: 1)
+
+                if components.count == 2 {
+                    self.major = Int(components[0])
+                    self.minor = components[1].split(separator: ".").first.map(String.init).flatMap(Int.init)
+                } else {
+                    self.major = nil
+                    self.minor = nil
+                }
+            }
+        }
+
+        init(utsname info: utsname) {
+            self.machine = withUnsafePointer(to: info.machine) { pointer in
+                pointer.withMemoryRebound(to: UInt8.self, capacity: MemoryLayout.size(ofValue: pointer)) {
```

Comment:
This doesn't look like this can possibly be right: the size of value seems absolutely wrong. Can I suggest we use `withUnsafeBytesOf` instead?